edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+
+        async with httpx.AsyncClient() as client:
+            resp = await client.post(
+                "https://github.com/login/oauth/access_token",
+                json=data,
+                headers=headers,
+            )
+            token = resp.json()['access_token']
+
+            return token
+
+
+providers = {
+    "github": GitHubProvider,
+}
+
+
+def _get_provider(name, client_id, client_secret):
+    provider_class = providers.get(name)
+    if provider_class is None:
+        raise errors.InternalServerError(f"unknown provider: {name}")
+    return provider_class(client_id, client_secret)
+
+
+def _get_auth_signing_key(db_config: SettingsMap):
+    auth_signing_key = _maybe_get_config(db_config, "xxx_auth_signing_key")
+    if auth_signing_key is None:
+        raise errors.InternalServerError(
+            "No auth signing key configured: Please set `cfg::Config.xxx_auth_signing_key`"
+        )
+    key_bytes = base64.b64encode(auth_signing_key.encode())
+
+    return jwk.JWK(kty="oct", k=key_bytes.decode())
+
+
+def _get_client_credentials(
+    provider_name: str, db_config: SettingsMap
+) -> tuple[str, str]:
+    client_id = _maybe_get_config(db_config, f"xxx_{provider_name}_client_id")
+    client_secret = _maybe_get_config(
+        db_config, f"xxx_{provider_name}_client_secret"
+    )
+    if client_id is None or client_secret is None:
+        raise errors.InternalServerError(
+            f"No client credentials configured for provider `{provider_name}`: "
+            f"Please set `cfg::Config.{provider_name}_client_id` and "
+            f"`cfg::Config.{provider_name}_client_secret`"
+        )
+    return (client_id, client_secret)
+
+
+T = TypeVar("T")
+
+
+def _maybe_get_config(
+    db_config: SettingsMap, key: str, expected_type: Type[T] = str
+) -> T | None:
+    value = db_config.get(key, (None, None, None, None))[1]
+
+    if value is None:
+        return None
+
+    if not isinstance(value, expected_type):
+        raise TypeError(
+            f"Config value `{key}` must be {expected_type.__name__}, got {type(value).__name__}"
+        )
+
+    return value
+
+
+def _make_signed_token(iss: str, provider: str, key: jwk.JWK) -> jwt.JWT:
+    expires_at = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
+
+    state_claims = {
+        "iss": iss,
+        "provider": provider,
+        "exp": expires_at.astimezone().timestamp(),
+    }
+    state_token = jwt.JWT(
+        header={"alg": "HS256"},
+        claims=state_claims,
+    )
+    state_token.make_signed_token(key)
+    return state_token
+
+
+def _get_search_param(query: str, key: str) -> str | None:
+    return urllib.parse.parse_qs(query).get(key, [None])[0]
+
+
+def _get_extension_path(url: httptools.parser.url_parser.URL) -> str:
+    path_parts = url.path.decode().split('/')
+
+    path_seg_parts = ["ext", "auth"]
+
+    try:
+        index = path_parts.index(path_seg_parts[0])
```

Comment:
this stuff seems fragile. Is there a better simpler way to build the URL?

Reply:
Yeah, I'm curious about this: there is nothing I could come up with that would know enough about the routing to do this consistently. I could get to this a little easier from the `protocol.pyx` code since we're parsing the request there and have more structured details about the route at that point. wdyt?